[chore] [processor/k8sattributes] improve documentation for the k8sattributes processor (#33800)
open-telemetry/opentelemetry-collector-contrib: processor/k8sattributesprocessor/README.md

to the pod from where the datapoint originated, so we can add the relevant pod metadata to the datapoint. By default, it associates the incoming connection IP
to the Pod IP. But for cases where this approach doesn't work (sending through a proxy, etc.), a custom association rule can be specified.

Each association is specified as a list of sources of associations. The maximum number of sources within an association is 4. 
A source is a rule that matches metadata from the datapoint to pod metadata.
In order to get an association applied, all the sources specified need to match.

Each sources rule is specified as a pair of `from` (representing the rule type) and `name` (representing the attribute name if `from` is set to `resource_attribute`).
The following rule types are available:

  - `connection`: Takes the IP attribute from connection context (if available). In this case the processor must appear before any batching or tail sampling, which remove this information.
  - `resource_attribute`: Allows specifying the attribute name to lookup in the list of attributes of the received Resource. Semantic convention should be used for naming.

>    Available attribute names for association sources are the following:

Author: please let me know if this list is incomplete or wrong, but I thought it might be worth having this explicit list to avoid confusion regarding which attributes can be used for these rules (e.g. the limitation of container level attributes not working here)

    - k8s.namespace.name
    - k8s.pod.name
    - k8s.pod.hostname
    - k8s.pod.ip
    - k8s.pod.start_time
    - k8s.pod.uid

>     - k8s.replicaset.uid

Reviewer: Shouldn't these be either enabled by default or explicitly set by the `metadata` rules in order to be available in the association? If so, I think it makes sense to mention that.

Author: good point, I have moved the list to the section about extending the metadata and added a remark that they need to be set there to be available in the association

   instance. If it's not set, the latest container instance will be used:
   - container.id (not added by default, has to be specified in `metadata`)

Please note however that container level attributes can not be used for source rules in the pod_association.

Example for extracting container level attributes:

> ```yaml

Author: may be a trivial example, but it's always helpful especially for new users to see exactly where these properties need to be set. Can remove it though if it makes the readme too verbose

> extract:

Reviewer: While this example is useful, I think we lack an example showing what the output of the processor looks like given specific inputs.

Author: you're right - having an example here should make things clearer - I have added a small example to illustrate how the attributes of a span are extended based on the properties of a matching pod

   instance. If it's not set, the latest container instance will be used:
   - container.id (not added by default, has to be specified in `metadata`)

> Please note however that container level attributes can not be used for source rules in the pod_association.

Reviewer 2: ```suggestion
Please note, however, that container level attributes can't be used for source rules in the pod_association.
```

Example for extracting container level attributes:

```yaml
pod_association:
- sources:
    - from: connection
extract:
  metadata:
  - k8s.pod.name
  - k8s.pod.uid
  - container.image.name
  - container.image.tag
  - k8s.container.name
```

> This configuration will attach the attributes listed in the `metadata` section to all resources received by a matching pod
> with the `k8s.container.name` attribute being present. For example, given the following trace

Reviewer 2: ```suggestion
The previous configuration attaches the attributes listed in the `metadata` section to all resources received by a matching pod with the `k8s.container.name` attribute being present. For example, when the following trace
```

```json
{
  "name": "lets-go",
  "context": {
    "trace_id": "0x5b8aa5a2d2c872e8321cf37308d69df2",
    "span_id": "0x051581bf3cb55c13"
  },
  "parent_id": null,
  "start_time": "2022-04-29T18:52:58.114201Z",
  "end_time": "2022-04-29T18:52:58.114687Z",
  "attributes": {
    "k8s.container.name": "telemetrygen"
  }
}
```

is sent to the collector by the following pod,

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    workload: deployment
  name: telemetrygen-pod
  namespace: e2ek8senrichment
  uid: 038e2267-b473-489b-b48c-46bafdb852eb
spec:
  containers:
  - command:
    - /telemetrygen
    - traces
    - --otlp-insecure
    - --otlp-endpoint=otelcollector.svc.cluster.local:4317
    - --duration=10s
    - --rate=1
    - --otlp-attributes=k8s.container.name="telemetrygen"
    image: ghcr.io/open-telemetry/opentelemetry-collector-contrib/telemetrygen:latest
    name: telemetrygen
status:
  podIP: 10.244.0.11
```

> the processor will associate the received trace to the pod, based on the connection ip, and add those attributes to the resulting span:

Reviewer 2: ```suggestion
the processor associates the received trace to the pod, based on the connection IP, and add those attributes to the resulting span:
```
